Mage.Common/src/main/java/mage/view/CardView.java

```diff
@@ -745,13 +745,7 @@ private void generateCardIconsForAny(MageObject object, Ability ability, Game ga
                 && showAbility != null
                 && (showZone.match(Zone.BATTLEFIELD) || showZone.match(Zone.STACK))) {
             int costX;
-            if (showCard instanceof Permanent) {
-                // permanent on battlefield (can show x icon multiple turns, so use end_game source)
```

Comment:
Is it works fine and show x forever on permanent?

Reply:
Yes, the two paths are already executing the same code and have been since the Costs Tags rework.